### PR TITLE
Add spawner and unspawner scripts

### PR DESCRIPTION
**Pull request by v-lopez**
Fixes https://github.com/ros-controls/ros2_control/issues/294

### CODE REVIEW

**Comment by v-lopez** [comment]
This works better if you have https://github.com/ros-controls/ros2_control/pull/309 to detect when a command failed.

And also https://github.com/ros2/ros2cli/pull/590 for the param load functionality.

File: controller_manager/CMakeLists.txt
Comment (destogl) [inline controller_manager/CMakeLists.txt:66]:
```suggestion
        DESTINATION lib/${PROJECT_NAME}
)
```

File: controller_manager/scripts/unspawner.py
Comment (destogl) [inline controller_manager/scripts/unspawner.py:37]: Is this comment related to stopping controller? Can you group it there? It is confusing as it is now.

**Review by destogl — CHANGES_REQUESTED** [review]
Request for 2 small changes. 

Were spawner and unspawner nodes in ros1. Does it make sense to have them as nodes or just as scripts as you implemented?

**Comment by v-lopez** [comment]
> Request for 2 small changes.
> 
> Were spawner and unspawner nodes in ros1. Does it make sense to have them as nodes or just as scripts as you implemented?

That's up for debate. I started as a node, but I would need to duplicate much code from ros2controlcli and from the ros2param load. I opened https://github.com/ros2/rclpy/issues/671 and went to work as a script to above duplication.

If we want to add more features like listening to a topic as in ROS1, it makes more sense as a node, hopefully there's a way of avoiding duplicating everything.

**Comment by v-lopez** [comment]
I'd say go with the current version so we have something, and later when we can avoid code duplication, we rewrite it to use services directly.

One thing we could do now as a middleground solution, is implement it as a Node that calls the subprocess.
The reasoning being that if we want to write launch files using it, right now they would be written with [ExecuteProcess Action](https://github.com/ros2/launch/blob/master/launch/launch/actions/execute_process.py) and if it was a Node with the [Node Action](https://github.com/ros2/launch_ros/blob/master/launch_ros/launch_ros/actions/node.py), if we want launch files to remain unchanged, we should use a Node right now, even if it's implemented as a script.

File: controller_manager/scripts/unspawner.py
Comment (v-lopez) [reply controller_manager/scripts/unspawner.py]: done

File: controller_manager/CMakeLists.txt
Comment (v-lopez) [reply controller_manager/CMakeLists.txt]: done

**Comment by destogl** [comment]
> I'd say go with the current version so we have something, and later when we can avoid code duplication, we rewrite it to use services directly.

OK, if you correct flake8 I agree :)

**Comment by v-lopez** [comment]
Damn it, yeah I will.

Let me change it to node first (albeit using subprocess), I'll do it Friday morning.

**Comment by v-lopez** [comment]
I will, sorry about that.

Do you know why my colcon test and ament_flake8 won't detect that stuff? 

```
$ dpkg -l | grep flake8
ii  ros-foxy-ament-flake8                           0.9.6-1focal.20201210.043943         amd64        The ability to check code for style and syntax conventions with flake8.

$ ament_flake8  
2 files checked
No problems found

Checked files:

* ./spawner.py
* ./unspawner.py
```

**Comment by bmagyar** [comment]
I'm happy to merge this if you pinky promise a follow-up PR with some tests?

**Comment by v-lopez** [comment]
I will add the tests, but we may want to hold the merge until https://github.com/ros2/ros2cli/pull/590 is passed. Otherwise the part about loading parameters is going to fail because `ros2 param load` will not exist.

**Comment by bmagyar** [comment]
Is that going into Foxy or only Galactic?

**Comment by v-lopez** [comment]
I'm gonna request for a cherry-pick to foxy since it's 100% backwards compatible and useful.

**Comment by v-lopez** [comment]
I added the tests but they seem to fail on the buildfarm, I'll check them next week.

File: controller_manager/scripts/spawner.py
Comment (bmagyar) [inline controller_manager/scripts/spawner.py:23]:
```suggestion
from rclpy.duration import Duration
from rclpy.node import Node
```

**Comment by destogl** [comment]
Could you add an example of using spawners in launch files? Maybe on `ros2_control_demos` repo?

**Comment by v-lopez** [comment]
I'll add the examples.

Do you have any idea why is it failing?
I see some "No executable found" messages as if the spawner.py was not in the `ros2 run` path, but if I run the tests with `colcon test` it works for me.

Also I expect that they will fail until https://github.com/ros2/ros2cli/pull/596 is accepted, but they should fail differently complaining with:
>usage: ros2 param [-h] Call `ros2 param <command> -h` for more detailed usage. ...
ros2 param: error: argument Call `ros2 param <command> -h` for more detailed usage.: invalid choice: 'load' (choose from 'delete', 'describe', 'dump', 'get', 'list', 'set')

**Comment by v-lopez** [comment]
@destogl @bmagyar to add the ros2_control_demo example I need to solve https://github.com/ros-controls/ros2_control/issues/320 first. To put the reusable code there. It's 70 lines of python launch nonsense that I don't want to replicate.

**Comment by destogl** [comment]
> @destogl @bmagyar to add the ros2_control_demo example I need to solve #320 first. To put the reusable code there. It's 70 lines of python launch nonsense that I don't want to replicate.

fully agree with you!

File: controller_manager/controller_manager/launch_utils.py
Comment (destogl) [inline controller_manager/controller_manager/launch_utils.py:67]: Typo

```suggestion
    # Setting --unload-on-kill if launch arg unload_on_kill is "true"
```

File: controller_manager/package.xml
Comment (destogl) [inline controller_manager/package.xml:16]: Since you are adding many dependencies here. Does it make sense to move `launch_utils.py` into a new package?

File: controller_manager/scripts/spawner.py
Comment (destogl) [inline controller_manager/scripts/spawner.py:67]: nitpick

```suggestion
```

File: controller_manager/scripts/spawner.py
Comment (destogl) [inline controller_manager/scripts/spawner.py:83]: Does it make sense to run this if the controller is already loaded? If not can you move this a few lines later in or after the `else:`-part?

File: controller_manager/scripts/spawner.py
Comment (destogl) [inline controller_manager/scripts/spawner.py:97]: will this also be executed if the controller is loaded? I this intended?

File: controller_manager/scripts/spawner.py
Comment (destogl) [inline controller_manager/scripts/spawner.py:115]: This could clutter the output. Maybe to throttle this to 60s or something?

**Review by destogl — COMMENTED** [review]
I like this! Very nice work! Here few comments/proposals.

File: controller_manager/package.xml
Comment (v-lopez) [reply controller_manager/package.xml:16]: Maybe, although they are quite basic dependencies. 

I put it here because the original spawner was in this repo. But it could be somewhere else.

@bmagyar thoughts?

File: controller_manager/scripts/spawner.py
Comment (v-lopez) [reply controller_manager/scripts/spawner.py:115]: It shouldn't, the print is before the loop. Will only be printed once.

File: controller_manager/scripts/spawner.py
Comment (v-lopez) [reply controller_manager/scripts/spawner.py:97]: Yes, I believe that if you are calling this script you expect that at the end of it you have a loaded + configured + started controller, and if you provided a parameter file, it should be loaded. 

So when you are testing parameter, you would stop the controller and call this script to load params, configure and start.

File: controller_manager/scripts/spawner.py
Comment (v-lopez) [reply controller_manager/scripts/spawner.py:83]: You are right, it doesn't make sense if the controller is loaded.

Moving it below, thanks!

**Comment by v-lopez** [comment]
@Karsten1987 This is the PR we were talking about. Waiting on ros2param to be released.

File: controller_manager/CMakeLists.txt
Comment (Karsten1987) [inline controller_manager/CMakeLists.txt:67]:
```suggestion
install(PROGRAMS
  scripts/spawner.py
  scripts/unspawner.py
  DESTINATION lib/${PROJECT_NAME}
)
```

File: controller_manager/package.xml
Comment (Karsten1987) [reply controller_manager/package.xml:16]: not speaking for Bence here, but I think it makes sense to have the spawner and its dependencies live in the controller manager repo. Given that the launch components are rather independent from the rest of the code, we can factor this out into its own package later on quite easily.

File: controller_manager/package.xml
Comment (Karsten1987) [inline controller_manager/package.xml:21]: alphasort

```suggestion
  <depend>ros2param</depend>
  <depend>ros2run</depend>
```

File: controller_manager/scripts/spawner.py
Comment (Karsten1987) [inline controller_manager/scripts/spawner.py:29]: I am not a big fan of spawning a subprocess here. This adds another dependency here (`ros2control`) which is unnecessary. Also, wouldn't you create a circular dependency here?

Why not directly call the controller manager service?

File: controller_manager/scripts/spawner.py
Comment (Karsten1987) [inline controller_manager/scripts/spawner.py:51]: Force the file to exist.

```suggestion
        '-p', '--param-file', type=argparse.FileType('r'), help='Controller param file to be loaded into controller '
```

File: controller_manager/scripts/spawner.py
Comment (Karsten1987) [inline controller_manager/scripts/spawner.py:76]: Same comments as above, why not calling `node.get_node_names` and verify that the controller manager is in there?
example: https://github.com/ros2/ros2cli/blob/master/ros2node/ros2node/api/__init__.py#L60

File: controller_manager/scripts/spawner.py
Comment (Karsten1987) [inline controller_manager/scripts/spawner.py:93]: same, I feel there should be python API for this.

File: controller_manager/package.xml
Comment (Karsten1987) [inline controller_manager/package.xml:19]: this should only be a `exec_depend`

**Review by Karsten1987 — CHANGES_REQUESTED** [review]
I have two requests for changes here:

* Replace the python subprocess calls with actual ROS2 python API. Most of the calls can easily be replaced by a service call. That saves error prone command line output parsing.
* I'd love to see this functionality refactored into some sort of spawner API. The reason for this is that I believe we should leverage this quite a bit within the ros2 control CLI, i.e. loading a controller or unloading it.

File: controller_manager/scripts/spawner.py
Comment (v-lopez) [reply controller_manager/scripts/spawner.py:29]: The dependency is `ros2controlcli` which would not make a circular dependency.

I went for the subprocess approach because it was faster, but also to avoid duplicating the `ros2controlcli` code and also because loading parameter files is not trivial, there's no standard api for it (https://github.com/ros2/rclpy/issues/671) so I feel like we'd still keep using a subprocess for `ros2 param load` to avoid duplicating all that.

@destogl also mentioned using the services directly, my proposal was to use this to have something ready quick, but it turns out that we're still one month away from the sync that includes `ros2 param load` so we may have time.

On the other hand, I'm going to disappear for some months any time now (paternity leave) and I'd like to not find this waiting here when I'm back, so if I can't do it the next few days I'd ask to either someone give it a shot to change it, or accept as it is and improve it later.

**Review by destogl — APPROVED** [review]

File: controller_manager/scripts/spawner.py
Comment (v-lopez) [reply controller_manager/scripts/spawner.py:51]: Thanks, didn't know about this.

File: controller_manager/scripts/spawner.py
Comment (v-lopez) [reply controller_manager/scripts/spawner.py:76]: Initially I did that, but if you launch this together with gazebo there's some race condition between the startup of the spawner and the controller manager. 
When this is modified to use the services directly, it should wait for the service to exist using the rclpy api I believe.

File: controller_manager/scripts/spawner.py
Comment (v-lopez) [reply controller_manager/scripts/spawner.py:29]: https://github.com/ros-controls/ros2_control/issues/344

**Comment by bmagyar** [comment]
I have rebased the branch to resolve the conflict. We are still blocked waiting for a release of https://github.com/ros2/ros2cli/pull/596 , `ros2cli` to Foxy.

**Pull request merged**
